assets/js/components/read-more/index.js

```diff
@@ -0,0 +1,167 @@
+
+		this.state = {
+			isExpanded: true,
+			isClamped: false,
+			content: '.',
+		};
+
+		this.lineHeight = 0;
+		this.start = 0;
+		this.middle = 0;
+		this.end = 0;
+		this.originalContent = 1 === React.Children.count( props.children ) ? props.children : props.children.join( '' );
+
+		this.element = createRef();
+		this.getButton = this.getButton.bind( this );
+		this.onClick = this.onClick.bind( this );
+		this.clampLines = this.clampLines.bind( this );
+		this.moveMarkers = this.moveMarkers.bind( this );
+	}
+
+	componentDidMount() {
+		if ( this.props.children ) {
+			this.lineHeight = this.element.current.clientHeight + 1;
+			this.clampLines();
+		}
+	}
+
+	/**
+	 * Clamp lines calculates the height of a line of text and then limits it to the
+	 * value of the lines prop. Content is updated once limited.
+	 */
+	clampLines() {
+		const { ellipsis, maxLines } = this.props;
+		const maxHeight = ( this.lineHeight * maxLines ) + 1;
+
+		this.start = 0;
+		this.middle = 0;
+		this.end = this.originalContent.length;
+
+		while ( this.start <= this.end ) {
+			this.middle = Math.floor( ( this.start + this.end ) / 2 );
+			this.element.current.innerHTML = this.originalContent.slice( 0, this.middle );
+
+			if ( this.middle === this.originalContent.length ) {
+				this.setState( {
+					content: this.originalContent,
+					isClamped: false,
+				} );
+				return;
+			}
+
+			this.moveMarkers( maxHeight );
+		}
+
+		this.element.current.innerHTML = this.originalContent.slice( 0, this.middle - 5 ) + ellipsis;
+
+		this.setState( {
+			content: this.originalContent.slice( 0, this.middle - 5 ) + ellipsis,
+			isExpanded: false,
+			isClamped: true,
+		} );
+	}
+
+	moveMarkers( maxHeight ) {
+		if ( this.element.current.clientHeight <= maxHeight ) {
+			this.start = this.middle + 1;
+		} else {
+			this.end = this.middle - 1;
+		}
+	}
+
+	getButton() {
+		const { isExpanded, isClamped } = this.state;
+		const { className, lessText, moreText } = this.props;
+
+		const buttonText = isExpanded ? lessText : moreText;
+
+		if ( ! buttonText || ! isClamped ) {
+			return;
+		}
+
+		return (
+			<a
+				href="#more"
+				className={ className + '__read_more' }
+				onClick={ this.onClick }
+				aria-expanded={ ! isExpanded }
+			>
+				{ buttonText }
+			</a>
+		);
+	}
+
+	/**
+	 * Handles the click event for the read more/less button.
+	 *
+	 * @param {obj} e event
+	 */
+	onClick( e ) {
+		const { isExpanded } = this.state;
+
+		e.preventDefault();
+
+		if ( isExpanded ) {
+			this.clampLines();
+		} else {
+			this.setState( {
+				content: this.originalContent,
+			} );
+		}
+
+		this.setState( { isExpanded: ! isExpanded } );
+	}
+
+	render() {
+		const { className } = this.props;
+		const { content } = this.state;
+
+		if ( ! content ) {
+			return null;
+		}
+
+		return (
+			<div className={ className }>
+				<div
+					ref={ this.element }
+					dangerouslySetInnerHTML={ {
+						// You must ensure HTML content is safe for direct rendering.
```

Comment:
I'm not sure to understand this comment.

Should we keep the previous one?
```suggestion
						// `content` is the `review` parameter returned by the `reviews` endpoint.
						// It's filtered with `wp_filter_post_kses()`, which removes dangerous HTML tags,
						// so using it inside `dangerouslySetInnerHTML` is safe.
```